src/fs_terminal_commands.c

```diff
@@ -0,0 +1,244 @@
+    } else {    // Relative path
+        strcpy(abs_path, cur_path);
+        next_dir = strtok(path_copy, "/");
+    }
+
+    while (next_dir) {
+        process_path_part(next_dir, abs_path);
+        if (!abs_path[0]) {     // An error occurred on this part of the path
+            return;
+        }
+        next_dir = strtok(0, "/");
+    }
+}
+
+/**
+ * @brief Lists contents of a single directory
+ * @param dname Name of directory to list
+ */
+void ls_dir(const char *dname) {
+    char abs_path[256];
+    get_abs_path(dname, abs_path);
+    struct iso_dir *dir = iso_dopen(abs_path, 3);
+    if (dir) {
+        struct directory_record *dr = iso_dread(dir);
+        while (dr) {
+            if (is_dir(dr->file_flags[0])) {
+                console_set_fgcolor(100,100,255);
+            } else {
+                console_set_fgcolor(0,255,0);
+            }
+            console_printf("%s\n", dr->file_identifier);
+            dr = iso_dread(dir);
+        }
+        console_set_fgcolor(255,255,255);
+        iso_dclose(dir);
+    } else {
+        console_printf("ls: no directory %s\n", dname);
+    }
+    iso_dclose(dir);
+}
+
+/**
+ * @brief Move into a directory below yourself
+ * @details Move into the directory below yourself specified by dname
+ * @param dname Name of directory into which you attempt to move
+ */
+void move_into_directory(const char *dname, char *abs_path) {
+    if (strcmp(abs_path, ROOT) != 0) {   // Not at root, so we need to add a '/' to the path
+        int abs_path_len = strlen(abs_path);
+        memcpy(abs_path + abs_path_len, "/", 1);
+        memcpy(abs_path + abs_path_len + 1, dname, strlen(dname) + 1);
+    } else {
+        memcpy(abs_path + strlen(abs_path), dname, strlen(dname) + 1);
+    }
+}
+
+/**
+ * @brief Moves up one directory
+ * @details Changes working directory to the parent of current directory
+ */
+void move_up_directory(char *abs_path) {
+    if (strcmp(abs_path, "/") == 0) {
+        abs_path[0] = '\0';
+        return;
+    }
+    int i;
+    // Move up one directory by finding the last occurrence of '/'
+    for (i = strlen(abs_path); i >= 0; i--) {
+        if (i == 0) {
+            strcpy(abs_path, ROOT);
+            return;
+        }
+        else if (abs_path[i] == '/') {
+            abs_path[i] = '\0';
+            return;
+        }
+    }
+}
+
+/**
+ * @brief Moves up or down one directory
+ * @details Helper function of cmd_line_cd
+ * Moves up or down one directory, specified by dname
+ * @param dname Next directory in which to move
+ */
+void process_path_part(const char *dname, char *abs_path) {
+    if (strcmp(dname, ".") == 0) {
+        return;
+    } else if (strcmp(dname, "..") == 0) {     // Move up one directory
+        move_up_directory(abs_path);
+        return;
+    } else {
+        move_into_directory(dname, abs_path);
+        return;
```

Comment:
I like it for the sake of clarity in control flow. If it's glaringly painful to look at, we can remove it.